parser/parser_test.go

```diff
@@ -639,22 +639,123 @@ func TestParseValidInput(t *testing.T) {
 			t.Errorf("test case: %s", testCase.input)
 			t.Errorf("length mismatch, expected %d definitions, got %d",
 				len(testCase.defs), len(file.Defs))
-		}
 
-		if len(file.Comments) == len(testCase.comments) {
-			for i := range file.Comments {
-				if !reflect.DeepEqual(file.Comments[i], testCase.comments[i]) {
-					t.Errorf("test case: %s", testCase.input)
-					t.Errorf("incorrect comment %d:", i)
-					t.Errorf("  expected: %s", testCase.comments[i])
-					t.Errorf("       got: %s", file.Comments[i])
-				}
-			}
-		} else {
-			t.Errorf("test case: %s", testCase.input)
-			t.Errorf("length mismatch, expected %d comments, got %d",
-				len(testCase.comments), len(file.Comments))
+// Assuming that the two trees have equivalent structure, this function returns a map that translates nodes in <actual> to <expected>
+// The behavior is undefined if the two trees don't have equal structure between their non-comment nodes
+func getNodeMapping(expected *SyntaxTree, actual *SyntaxTree) (remappings map[ParseNode]ParseNode) {
+	actualNodes := actual.ListOfAllNodes()
+	expectedNodes := expected.ListOfAllNodes()
+	remappings = make(map[ParseNode]ParseNode, 0)
+	for i, actual := range actualNodes {
+		remappings[actual] = expectedNodes[i]
+	}
+	return remappings
+}
+
+// Tells whether the attached comments are equivalent between the two trees
+func compareComments(expected *ParseTree, actual *ParseTree, actualNode_remappings map[ParseNode]ParseNode) proptools.Comparison {
+	// Construct a new map of comments, whose values are the same as <actual>, but whose keys are the corresponding keys in <expected>
+	// The reason we replace the keys is so proptools.DeepCompare can know which keys we want to compare against which others
+
+	analogousActualComments := make(map[ParseNode](*CommentPair), 0)
+	for actualKey, actualValue := range actual.SyntaxTree.comments {
+		analogousKey := actualNode_remappings[actualKey]
+		analogousActualComments[analogousKey] = actualValue
+	}
+
+	// now compare the comments
+	return proptools.DeepCompare(
+		proptools.NamedItem{"expected.SyntaxTree.Comments", expected.SyntaxTree.comments},
+		proptools.NamedItem{"actual.SyntaxTree.Comments", analogousActualComments})
+}
+
+func compareSourcePositions(expected *ParseTree, actual *ParseTree, actualNode_remappings map[ParseNode]ParseNode) proptools.Comparison {
```

Comment:
Sometimes its better to use a short, readable name and put the extra information in a comment, rather than carrying all the information everywhere you use the name.  See https://github.com/golang/go/wiki/CodeReviewComments#variable-names